File: rust-native/examples/smoke.rs
use rustwright::{chromium, ActionOptions, GotoOptions, LaunchOptions, ScreenshotOptions};
use serde_json::json;

fn main() -> rustwright::Result<()> {
    let html = r#"<!doctype html><html><head><title>Rustwright Rust Smoke</title></head><body><h1 id="message">ready</h1><input id="name"><button id="go" onclick="document.querySelector('#message').textContent=document.querySelector('#name').value">Go</button></body></html>"#;
    let url = format!("data:text/html;charset=utf-8,{}", percent_encode(html));
    let browser = chromium().launch(LaunchOptions::default())?;
    let result = (|| {
        let page = browser.new_page()?;
        page.goto(&url, GotoOptions::default())?;
        let title = page.title(ActionOptions::default())?;
        let before = page.text_content("#message", ActionOptions::default())?;
        page.fill("#name", "Rustwright for Rust", ActionOptions::default())?;
        page.click("#go", ActionOptions::default())?;
        let after = page.text_content("#message", ActionOptions::default())?;
        let value = page.evaluate(
            "document.querySelector('#name').value",
            None,
            ActionOptions::default(),
        )?;
        let screenshot = page.screenshot(ScreenshotOptions::default())?;
        println!(
            "{}",
            json!({
                "title": title,
                "before": before,
                "after": after,
                "value": value,
                "screenshotBytes": screenshot.len(),
            })
        );
        page.close(Default::default())
    })();
    let close_result = browser.close();
    result.and(close_result)
}

fn percent_encode(value: &str) -> String {
    let mut encoded = String::with_capacity(value.len());
    for byte in value.bytes() {
        if byte.is_ascii_alphanumeric() || matches!(byte, b'-' | b'.' | b'_' | b'~') {
            encoded.push(byte as char);
        } else {
            encoded.push_str(&format!("%{byte:02X}"));
        }
    }
    encoded
}

File: node/smoke.mjs
import { join } from 'node:path';
import { tmpdir } from 'node:os';
import { chromium } from './index.mjs';

const screenshotPath = join(tmpdir(), `rustwright-node-smoke-${process.pid}.png`);
const html = `<!doctype html>
<html>
  <head><title>Rustwright Node Smoke</title></head>
  <body>
    <h1 id="message">ready</h1>
    <input id="name" />
    <button id="go" onclick="document.querySelector('#message').textContent = document.querySelector('#name').value">Go</button>
  </body>
</html>`;

const browser = await chromium.launch({ headless: true });
try {
  // Exercise the BrowserContext shim and page-level default timeouts alongside
  // the core page methods.
  const context = await browser.newContext();
  context.setDefaultTimeout(30_000);
  const page = await context.newPage();
  page.setDefaultNavigationTimeout(30_000);
  await page.goto(`data:text/html,${encodeURIComponent(html)}`);
  const title = await page.title();
  const before = await page.textContent('#message');
  await page.fill('#name', 'Rustwright for Node');
  await page.click('#go');
  const after = await page.textContent('#message');
  const value = await page.evaluate(() => document.querySelector('#name').value);
  const screenshot = await page.screenshot({ path: screenshotPath });
  const contextPages = context.pages().length;
  await context.close();
  console.log(JSON.stringify({
    title,
    before,
    after,
    value,
    contextPages,
    screenshotBytes: screenshot.length,
    screenshotPath
  }, null, 2));
} finally {
  await browser.close();
}
